experimental/casemapping/src/exceptions.rs

```diff
@@ -0,0 +1,394 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use yoke::{Yokeable, ZeroCopyFrom};
+use zerovec::{VarZeroVec, ZeroVec};
+
+use crate::error::Error;
+use crate::internals::{ClosureSet, DotType, MappingKind};
+
+// This represents case mapping exceptions that can't be represented as a delta applied to
+// the original code point. Similar to ICU4C, data is stored as a u16 array. The codepoint
+// trie in CaseMapping stores offsets into this array. The u16 at that index contains a
+// set of flags describing the subsequent data.
```

Comment:
Discussion: I've said this before verbally, but I'll put it in writing. It really would be a *lot* more elegant with regard to code readability, ease of validation, and consistency across ICU4X if you would make this `VarZeroVec<ExceptionItemULE>` with something along the lines of

```rust
#[derive(VarULE)]
struct ExceptionItemULE {
    header: /* this can be a structured header type */,
    replacements: str, // you can put your UTF-8 in here and elimiate the side table
}
```

I'm quite confident you could cut the size of this file, and probably some of the sister files, at least in half if you did it this way.

The *only* downside I'm aware of is that that this incurs an additional "hop" since we need to look at the index table.  I'd be interested to see if this is a big enough penalty to show up in benchmarks.

Another model that might solve the problem would be to store code point replacements in the side table along with string replacements.  You could then have `ZeroVec<ExceptionItemULE>` where `ExceptionItemULE` is a fixed-width type containing the indices into the string.  You then do not need any other lookups.